File: test/dummy/app/components/dummy/searches/show_component.rb
# frozen_string_literal: true

class Dummy::Searches::ShowComponent < ApplicationComponent
  include SimpleForm::ActionViewExtensions::FormHelper

  def initialize(search:)
    @search = search
  end

  def form(&block)
    opts = {
      url: controller.main_app.dummy_search_path,
      method: :get,
      html: {
        class: "d-searches-show__form h1",
        id: nil,
        data: stimulus_data(target: "form", action: "onFormSubmit")
      },
    }

    simple_form_for("", opts, &block)
  end

  def query_input(f)
    f.input :q,
            input_html: {
              class: "d-searches-show__input",
              value: params[:q],
              placeholder: t(".placeholder"),
              autofocus: true,
              autocomplete: "off",
              id: nil,
              data: stimulus_data(target: "input", action: { input: "onInputInput", focus: "onInputFocus" })
            },
            label: false,
            wrapper: false
  end

  def data
    stimulus_controller("d-searches-show", values: {
      loading: false,
    })
  end
end

File: test/dummy/app/components/dummy/searches/show_component.js
//= require folio/add_params_to_url
//= require folio/debounce

window.Folio.Stimulus.register('d-searches-show', class extends window.Stimulus.Controller {
  static targets = ['form', 'input', 'contentsWrap']

  static values = {
    loading: Boolean
  }

  connect () {
    this.moveInputCursor()
  }

  disconnect () {
    this.abortLoad()
  }

  debouncedLoad = window.Folio.debounce(() => {
    const paramsHash = {}

    if (this.inputTarget.value) {
      paramsHash.q = this.inputTarget.value
    }

    const urlParams = new URLSearchParams(window.location.href)
    const tab = urlParams.get('tab')
    if (tab) paramsHash.tab = tab

    const url = Object.keys(paramsHash).length ? window.Folio.addParamsToUrl(this.formTarget.action, paramsHash) : this.formTarget.action

    this.abortLoad()
    this.abortController = new AbortController()

    if (window.Turbolinks) {
      window.Turbolinks.controller.replaceHistoryWithLocationAndRestorationIdentifier(url, window.Turbolinks.uuid())
    }

    window.Folio.Api.apiGet(url, null, this.abortController.signal).then((res) => {
      if (res && res.data) {
        this.contentsWrapTarget.innerHTML = res.data
      } else {
        throw new Error('No data')
      }
    }).catch((err) => {
      if (err.name === 'AbortError') return
      window.Dummy.Ui.Flash.alert(err.message)
    }).finally(() => {
      this.loadingValue = false
    })
  })

  load () {
    if (!this.loadingValue) this.loadingValue = true
    this.debouncedLoad()
  }

  onFormSubmit (e) {
    e.preventDefault()
    this.load()
  }

  onInputFocus () {
    this.moveInputCursor()
  }

  moveInputCursor () {
    const length = this.inputTarget.value.length
    this.inputTarget.setSelectionRange(length, length)
  }

  onInputInput (e) {
    this.load()
  }

  abortLoad () {
    if (this.abortController) {
      this.abortController.abort()
      this.abortController = null
    }
  }
})
